draft-ietf-quic-recovery.md

```diff
@@ -767,20 +767,22 @@ While in slow start, QUIC increases the congestion window by the
 number of bytes acknowledged when each acknowledgment is processed, resulting
 in exponential growth of the congestion window.
 
-QUIC exits slow start upon loss or upon increase in the ECN-CE counter.
-When slow start is exited, the congestion window halves and the slow start
-threshold is set to the new congestion window.  QUIC re-enters slow start
-any time the congestion window is less than the slow start threshold,
-which only occurs after persistent congestion is declared.
+QUIC MUST exit slow start and enter congestion avoidance upon loss or upon
+increase in the ECN-CE counter. When slow start is exited, the congestion
+window halves and the slow start threshold is set to the new congestion
```

Comment:
Restating this here: With https://tools.ietf.org/html/rfc5681#section-4.3 ssthresh is reduced upon entering recover but cwnd is set to ssthresh when leaving recovery.

Reply:
That's what TCP does, but that's not necessary in QUIC. Consider what happens in TCP: until the sender leaves recovery, it cannot grow the cwnd, and since the left edge of the cwnd is anchored to the cumulative ack point, the sender cannot send anything new. It can only keep retransmitting. In other words, the cwnd governs how much data to send and also _which data to send_. To allow for a sender to send new data, TCP allows inflating the cwnd which is deflated on leaving recovery, which in the best case is 1 RTT after loss.

QUIC deliberately avoids this problem, and the cwnd is not anchored to a left edge. It's an abstract limit on the bytes in flight, and says nothing about which data can be sent. So, as data gets acked, and as the bytes in flight reduces to 1/2 the value at loss, the sender can start sending again. This would happen 1/2 RTT after loss.

So at the end of the "recovery period", i.e., 1 RTT after loss:
- TCP would have sent nothing if its cwnd were reduced immediately on loss. 
- QUIC would have been sending for 1/2 RTT (it would send as much as is allowed by the new cwnd basically).

This is also why simply referencing 5681 won't do. The semantics of cwnd are different.

Reply:
Let me start by saying that QUIC doesn't have to be precisely the same as TCP -- this is why we are not normatively referencing RFC 5681. Even if this is a behavior difference from deployed TCP, this is not a reduction in quality. What is the change you're proposing?

That said, what you say isn't true in TCP. Specifically, RFC 5681 says (in Section 4.3):
```
   That is, when the first loss in a window of data is detected,
   ssthresh MUST be set to no more than the value given by equation (4).
   Second, until all lost segments in the window of data in question are
   repaired, the number of segments transmitted in each RTT MUST be no
   more than half the number of outstanding segments when the loss was
   detected.
```
That last sentence is critical. With just RFC 5681, TCP will do exactly as QUIC does. This isn't true of Linux TCP however, which uses rate halving... the PRR document (RFC 6937, Section 1) goes through this is some detail, and is in fact one of the reasons for PRR's existence.

Reply:
This text from RFC5681 is fine and if you want we can take over that language. That text does not say that you need to immediately need to halve the cwnd but you can't transmit more than half the cwnd _over the whole RTT_. That's also what rate halving does actually but by reducing gradually over one RTT you avoid not sending anything for half an RTT and then send full rate in the second half (what the originally fast recovery did as nicely explain in section 1 of RFC6937). PRR addresses a different problem of rate halving which is that you actually might reduce too much (e.g. for cubic when you only want to reduce by 30%, or when multiple packets are lost). PRR still uses rate halving but then goes back into SS to make sure you reach your ssthresh quickly in case you have reduced too much. This is also what is specified for QUIC at the moment.

Reply:
I've changed the text here to reflect this suggestion.

Reply:
Looks good now! Thanks!